deepr/layers/base.py

```diff
@@ -336,103 +343,57 @@ def __init__(self, offset, n_in=1, n_out=1, inputs=None, outputs=None, name=None
-    2
-    >>> add(1, tf.estimator.ModeKeys.PREDICT)
-    1
-
-    Note that 'tensors' and 'mode' need to be the the first arguments
-    of the function IN THIS ORDER.
-
-    Another way of using the decorator is on functions that create layer
-    instances. This allows you to create factories of layers and make
-    them :class:`~Layer` classes.
-
-    For example, the following snippet defines a subclass of :class:`~Layer`
-    whose `forward` method is the same as the layer returned by the
-    function.
-
-    >>> @layer(n_in=1, n_out=1)
-    ... def AddOne() -> Layer:
-    ...     return AddOffset(offset=1)
-    >>> add_one = AddOne()
-    >>> add_one(1)
-    2
-
-    A nice feature of the :class:`~layer` decorator is laziness: the code
-    inside the function is not executed until a call on a tensor. In
-    other words:
-
-    >>> @layer(n_in=1, n_out=1)
-    ... def AddOffset(tensors, offset):
-    ...     print("Calling layer")
-    ...     return tensors + offset
-    >>> add_one = AddOffset(offset=1)
-    >>> add_one(1)
-    "Calling layer"
+    Note that you need to use either "tensors" or "tensors" and "mode"
+    as the first two parameters.
     """
     # pylint: disable=protected-access,invalid-name
-
-    def _create_layer_class(fn: Callable) -> Type[Layer]:
-        """Decorator that creates a Layer class from a function"""
+    def _create_layer_class(fn: Callable) -> Callable:
+        """Decorator that creates a Layer constructor."""
         parameters = inspect.signature(fn).parameters
         signature = inspect.Signature([param for key, param in parameters.items() if key not in {"tensors", "mode"}])
 
-        if "tensors" in parameters:  # From a `forward` function
-
-            if list(parameters.keys())[0] != "tensors":
-                raise TypeError(f"'tensors' should be the first positional argument")
-
-            if "mode" in parameters:
-
-                if list(parameters.keys())[1] != "mode":
-                    raise TypeError(f"'mode' should be the second positional argument")
-
-                def _forward(self, tensors, mode: str = None):
-                    return fn(tensors, mode, *self._args, **self._kwargs)
-
-            else:
-
-                def _forward(self, tensors, mode: str = None):
-                    # pylint: disable=unused-argument
-                    return fn(tensors, *self._args, **self._kwargs)
-
-        else:  # From a constructor
-
-            def _forward(self, tensors, mode: str = None):
-                return fn(*self._args, **self._kwargs).forward(tensors, mode)
-
+        # Check legacy use case
+        if "tensors" not in parameters:
+            msg = f"'tensors' should be the first parameter of {fn.__name__}. "
+            msg += "Using a decorator on a Layer constructor has been deprecated. "
+            msg += "The decorator will have no impact (please remove it)."
+            LOGGER.error(msg)
+            return fn
```

Comment:
Decorator works on the same cases as before, but on the constructor case, is not applied.
We can either keep it that way, or just fail (would prefer, that would force people to remove the useless decorators).